pandas/core/groupby/categorical.py

```diff
@@ -1,14 +1,72 @@
 from __future__ import annotations
 
+import dataclasses
+from typing import final
+
 import numpy as np
 
+from pandas._typing import npt
+
 from pandas.core.algorithms import unique1d
 from pandas.core.arrays.categorical import (
     Categorical,
     CategoricalDtype,
     recode_for_categories,
 )
-from pandas.core.indexes.api import CategoricalIndex
+from pandas.core.indexes.api import (
+    CategoricalIndex,
+    Index,
+)
+
+
+@final
+@dataclasses.dataclass
+class CategoricalGrouper:
+    original_grouping_vector: Categorical | None
+    new_grouping_vector: Categorical
+    observed: bool
+    sort: bool
+
+    def result_index(self, group_index: Index) -> Index:
+        if self.original_grouping_vector is None:
+            return group_index
+        assert isinstance(group_index, CategoricalIndex)
+        return recode_from_groupby(
+            self.original_grouping_vector, self.sort, group_index
+        )
+
+    @classmethod
+    def make(cls, grouping_vector, sort: bool, observed: bool) -> CategoricalGrouper:
+        new_grouping_vector, original_grouping_vector = recode_for_groupby(
```

Comment:
instead of saving new_grouping_vector as an instance variable, could just return it from this method, since it isn't ever needed by this class.

Reply:
why 'make' instead of just `__init__`?